diff --git a/sys/rump/net/lib/libshmif/if_shmem.c b/sys/rump/net/lib/libshmif/if_shmem.c
index 490ca82..90b4e50 100644
--- a/sys/rump/net/lib/libshmif/if_shmem.c
+++ b/sys/rump/net/lib/libshmif/if_shmem.c
@@ -161,7 +161,8 @@ allocif(int unit, struct shmif_sc **scp)
 	uint32_t randnum;
 	int error;
 
-	randnum = cprng_fast32();
+//	randnum = cprng_fast32();
+    randnum = 0xc282c830;
 	memcpy(&enaddr[2], &randnum, sizeof(randnum));
 
 	sc = kmem_zalloc(sizeof(*sc), KM_SLEEP);
@@ -225,7 +226,7 @@ initbackend(struct shmif_sc *sc, int memfd)
 	    && sc->sc_busmem->shm_magic != SHMIF_MAGIC) {
 		printf("bus is not magical");
 		rumpuser_unmap(sc->sc_busmem, BUSMEM_SIZE);
-		return ENOEXEC; 
+		return ENOEXEC;
 	}
 
 	/*
@@ -674,7 +675,7 @@ shmif_rcv(void *arg)
 		KASSERT(busmem->shm_gen >= sc->sc_devgen);
 
 		/* need more data? */
-		if (sc->sc_devgen == busmem->shm_gen && 
+		if (sc->sc_devgen == busmem->shm_gen &&
 		    shmif_nextpktoff(busmem, busmem->shm_last)
 		     == sc->sc_nextpacket) {
 			shmif_unlockbus(busmem);
diff --git a/sys/rump/net/lib/libshmif/shmif_busops.c b/sys/rump/net/lib/libshmif/shmif_busops.c
index bcf7182..e722051 100644
--- a/sys/rump/net/lib/libshmif/shmif_busops.c
+++ b/sys/rump/net/lib/libshmif/shmif_busops.c
@@ -111,7 +111,7 @@ shmif_buswrite(struct shmif_mem *busmem, uint32_t off, void *data, size_t len,
 
 	memcpy(busmem->shm_data + off, data, chunk);
 
-	DPRINTF(("buswrite: wrote %d bytes to %d", chunk, off));
+	DPRINTF(("buswrite: wrote %zd bytes to %d", chunk, off));
 
 	if (filledbus) {
 		*wrap = true;
@@ -122,7 +122,7 @@ shmif_buswrite(struct shmif_mem *busmem, uint32_t off, void *data, size_t len,
 		return (off + chunk) % BUSMEM_DATASIZE;
 	}
 
-	DPRINTF((", wrapped bytes %d to 0\n", len));
+	DPRINTF((", wrapped bytes %zd to 0\n", len));
 
 	shmif_advancefirst(busmem, 0, len);
 
diff --git a/sys/rump/net/lib/libshmif/shmifvar.h b/sys/rump/net/lib/libshmif/shmifvar.h
index 7e9e509..6b5ed1d 100644
--- a/sys/rump/net/lib/libshmif/shmifvar.h
+++ b/sys/rump/net/lib/libshmif/shmifvar.h
@@ -61,11 +61,12 @@ struct shmif_pkthdr {
 #define BUSMEM_SIZE (1024*1024)
 #define BUSMEM_DATASIZE (BUSMEM_SIZE - sizeof(struct shmif_mem))
 
-#if 0
+#if 1
 #ifdef _KERNEL
 #include <rump/rumpuser.h>
 #define DPRINTF(x) rumpuser_dprintf x
 #else
+#include <stdio.h>
 #define DPRINTF(x) printf x
 #endif
 #else
